piglot/solver/crate/solver.py

```diff
@@ -0,0 +1,198 @@
+"""Module for CRATE solver."""
+from typing import Dict, Any, Type
+import os
+import time
+import shutil
+import subprocess
+from multiprocessing.pool import ThreadPool as Pool
+import numpy as np
+from piglot.parameter import ParameterSet
+from piglot.solver.solver import Solver, Case, CaseResult, OutputField, OutputResult, InputData
+from piglot.solver.crate.fields import crate_fields_reader, CRATEInputData
+from piglot.utils.solver_utils import has_keyword, load_module_from_file
+
+
+class CRATESolver(Solver):
+    """CRATE solver."""
+
+    def __init__(
+            self,
+            cases: Dict[Case, Dict[str, OutputField]],
+            parameters: ParameterSet,
+            output_dir: str,
+            crate_bin: str,
+            parallel: int,
+            tmp_dir: str,
+            ) -> None:
+        """Constructor for the CRATE solver class.
+
+        Parameters
+        ----------
+        cases : Dict[Case, Dict[str, OutputField]]
+            Cases to be run and respective output fields.
+        parameters : ParameterSet
+            Parameter set for this problem.
+        output_dir : str
+            Path to the output directory.
+        crate_bin : str
+            Path to the main.py CRATE file.
+        parallel : int
+            Number of parallel processes to use.
+        tmp_dir : str
+            Path to the temporary directory.
+        """
+        super().__init__(cases, parameters, output_dir)
+        self.crate_bin = crate_bin
+        self.parallel = parallel
+        self.tmp_dir = tmp_dir
+
+    def _run_case(self, values: np.ndarray, case: Case, tmp_dir: str) -> CaseResult:
+        """Run a single case wth CRATE.
+
+        Parameters
+        ----------
+        values: np.ndarray
+            Current parameter values
+        case : Case
+            Case to run.
+        tmp_dir: str
+            Temporary directory to run the simulation
+
+        Returns
+        -------
+        CaseResult
+            Results for this case
+        """
+        # Copy input file replacing parameters by passed value
+        input_data = case.input_data.prepare(values, self.parameters, tmp_dir=tmp_dir)
+        input_file = input_data.input_file
+        case_name, _ = os.path.splitext(os.path.basename(input_file))
+        # Run CRATE (we don't use high precision timers here to keep track of the start time)
+        begin_time = time.time()
+        process_result = subprocess.run(
+            ['python3', self.crate_bin, input_file],
```

Comment:
Directly calling `python3` seems dangerous. Maybe we should add an optional option to specify the Python interpreter to use when calling CRATE.